include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -226,6 +227,70 @@ struct transform_reduce
             __local_mem[__local_idx] = __res;
         }
     }
+
+    template <typename _NDItemId, typename _Size, typename _AccLocal, typename... _Acc>
+    void
+    nonseq_impl(const _NDItemId __item_id, const _Size __n, const _Size __global_offset, const _AccLocal& __local_mem,
+                const _Acc&... __acc) const
+    {
+        auto __global_idx = __item_id.get_global_id(0);
+        auto __local_idx = __item_id.get_local_id(0);
+        const _Size __stride = __item_id.get_local_range(0);
+
+        const _Size __adjusted_global_id =
+            __global_offset + __item_id.get_group_linear_id() * __stride * __iters_per_work_item + __local_idx;
+        const _Size __adjusted_n = __global_offset + __n;
+
+        // Coalesced load and reduce from global memory
+        if (__adjusted_global_id + __stride * __iters_per_work_item < __adjusted_n)
+        {
+            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
+            _ONEDPL_PRAGMA_UNROLL
+            for (_Size __i = 1; __i < __iters_per_work_item; ++__i)
+                __res = __binary_op(__res, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            __local_mem[__local_idx] = __res;
+        }
+        else if (__adjusted_global_id < __adjusted_n)
+        {
+            // TODO: simplify to not use floats
+            const _Size __items_to_process =
+                std::max(((__adjusted_n - __adjusted_global_id - 1) / __stride) + 1, static_cast<_Size>(0));
```

Comment:
Can we use `__dpl_ceiling_div` here?

Reply:
In this case they are not equivalent, mainly if the numerator is cleanly divisible by the denominator we still want to go up 1 value. That does not happen with ceiling division. So I have kept the current maths.